Assignment2
-Yuntong Zhu（217221508）

In [11]:
import numpy as np
import pandas as pd
from scipy import stats
import seaborn as sns
from scipy.stats import ttest_ind
from scipy.stats import f_oneway

In [12]:
df = pd.read_csv("/Users/zoezhu/Downloads/experiment_dataset.csv")
df

Unnamed: 0  Age   Location   Device    Variant  Time Spent       CTR
0             0   62  Location2  Device2    Control   13.928669  0.084776
1             1   18  Location1  Device1  Variant B   11.310518  0.096859
2             2   21  Location2  Device1  Variant B   24.842100  0.097630
3             3   21  Location1  Device3  Variant B   20.061300  0.109783
4             4   57  Location1  Device2  Variant B   34.495503  0.068579
..          ...  ...        ...      ...        ...         ...       ...
995         995   39  Location2  Device2  Variant B   17.252030  0.092211
996         996   38  Location3  Device2    Control   30.075898  0.078151
997         997   60  Location2  Device3    Control   31.929223  0.125213
998         998   35  Location2  Device2  Variant B   14.680299  0.095423
999         999   55  Location1  Device1  Variant B   25.803169  0.166688

[1000 rows x 7 columns]

In [13]:
df.drop('Unnamed: 0', axis=1, inplace = True)

In [14]:
df.head()

Age   Location   Device    Variant  Time Spent       CTR
0   62  Location2  Device2    Control   13.928669  0.084776
1   18  Location1  Device1  Variant B   11.310518  0.096859
2   21  Location2  Device1  Variant B   24.842100  0.097630
3   21  Location1  Device3  Variant B   20.061300  0.109783
4   57  Location1  Device2  Variant B   34.495503  0.068579

In [15]:
# No missing data
df.isnull().sum()

Age           0
Location      0
Device        0
Variant       0
Time Spent    0
CTR           0
dtype: int64

In [43]:
# by location
df.groupby('Location')[['Time Spent','CTR']].mean()

Time Spent       CTR
Location                       
Location1   22.707286  0.110217
Location2   22.648998  0.108517
Location3   22.787691  0.108708

This looks mostly the same average result,use Anova test to get p-value.

In [17]:
# checking Age
loc_1 = df[df['Location'] == 'Location1']['Age']
loc_2 = df[df['Location'] == 'Location2']['Age']
loc_3 = df[df['Location'] == 'Location3']['Age']

p_val = stats.f_oneway(loc_1, loc_2, loc_3)
print(p_val)

F_onewayResult(statistic=2.156278540665205, pvalue=0.11629462727741363)


In [18]:
# checking Time Spent
loc_1 = df[df['Location'] == 'Location1']['Time Spent']
loc_2 = df[df['Location'] == 'Location2']['Time Spent']
loc_3 = df[df['Location'] == 'Location3']['Time Spent']

p_val = stats.f_oneway(loc_1, loc_2, loc_3)
print(p_val)

F_onewayResult(statistic=0.053578839670573676, pvalue=0.9478339402848069)


In [19]:
# checking CTR
loc_1 = df[df['Location'] == 'Location1']['CTR']
loc_2 = df[df['Location'] == 'Location2']['CTR']
loc_3 = df[df['Location'] == 'Location3']['CTR']

p_val = stats.f_oneway(loc_1, loc_2, loc_3)
print(p_val)

F_onewayResult(statistic=0.5792245145655729, pvalue=0.5605211716238133)


For these 3 locations, it seems that the users behave the same. They should have a high p-value cause the p-values of the three locations are similar, with no significant difference.

In [20]:
# check the same for device type.

In [44]:
# checking Age
Dev_1 = df[df['Device'] == 'Device1']['Age']
Dev_2 = df[df['Device'] == 'Device2']['Age']
Dev_3 = df[df['Device'] == 'Device3']['Age']

p_val = stats.f_oneway(Dev_1, Dev_2, Dev_3)
print(p_val)

F_onewayResult(statistic=0.8219183096017563, pvalue=0.4398852106386104)


In [46]:
# checking Time Spent
Dev_1 = df[df['Device'] == 'Device1']['Time Spent']
Dev_2 = df[df['Device'] == 'Device2']['Time Spent']
Dev_3 = df[df['Device'] == 'Device3']['Time Spent']

p_val = stats.f_oneway(Dev_1, Dev_2, Dev_3)
print(p_val)

F_onewayResult(statistic=0.26645371811833884, pvalue=0.7661459958744103)


In [45]:
# checking CTR
Dev_1 = df[df['Device'] == 'Device1']['CTR']
Dev_2 = df[df['Device'] == 'Device2']['CTR']
Dev_3 = df[df['Device'] == 'Device3']['CTR']

p_val = stats.f_oneway(Dev_1, Dev_2, Dev_3)
print(p_val)

F_onewayResult(statistic=0.7105872492654717, pvalue=0.4916042399968955)


The p-values are all greater than the commonly used significance level of 0.05. This means that we cannot reject the null hypothesis and conclude that there is no statistically significant difference in click-through rate or time spent between the groups.

Question1：Analyze the results to determine which feature (if any) results in CTR or Time Spent lift.

In [22]:
# Filter the data for Variant A
ctr_variant_a= df[df['Variant'] == 'Variant A']
ctr_variant_a

Age   Location   Device    Variant  Time Spent       CTR
7     39  Location2  Device1  Variant A   21.525217  0.149985
9     41  Location1  Device2  Variant A   27.642788  0.137266
13    30  Location2  Device3  Variant A   26.208502  0.087875
15    56  Location2  Device2  Variant A    7.800901  0.069781
17    41  Location2  Device3  Variant A   32.699437  0.114626
..   ...        ...      ...        ...         ...       ...
983   48  Location2  Device1  Variant A   28.171759  0.144396
984   20  Location1  Device3  Variant A   31.648129  0.121159
985   37  Location1  Device3  Variant A   24.543317  0.137414
992   44  Location1  Device2  Variant A   20.161483  0.098470
993   24  Location3  Device3  Variant A   27.390527  0.095718

[330 rows x 6 columns]

In [23]:
# The average CTR for Variant A
avg_ctr_variant_a = ctr_variant_a['CTR'].mean()
print("Average CTR for Variant A:", avg_ctr_variant_a)

Average CTR for Variant A: 0.12026949300288214


In [24]:
#Time Spent AVG & CTR for Varinat A
ctr_variant_a.mean()


/var/folders/fm/9pz_54hn7ynbrz17999wv77m0000gn/T/ipykernel_24386/3124725032.py:2: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  ctr_variant_a.mean()


Age           39.778788
Time Spent    24.805547
CTR            0.120269
dtype: float64

In [25]:
# Filter the data for Variant B
ctr_variant_b= df[df['Variant'] == 'Variant B']
ctr_variant_b

Age   Location   Device    Variant  Time Spent       CTR
1     18  Location1  Device1  Variant B   11.310518  0.096859
2     21  Location2  Device1  Variant B   24.842100  0.097630
3     21  Location1  Device3  Variant B   20.061300  0.109783
4     57  Location1  Device2  Variant B   34.495503  0.068579
5     27  Location3  Device1  Variant B   26.129246  0.149341
..   ...        ...      ...        ...         ...       ...
989   31  Location2  Device2  Variant B   32.952763  0.098476
991   35  Location3  Device1  Variant B   22.409808  0.100951
995   39  Location2  Device2  Variant B   17.252030  0.092211
998   35  Location2  Device2  Variant B   14.680299  0.095423
999   55  Location1  Device1  Variant B   25.803169  0.166688

[330 rows x 6 columns]

In [26]:
#the average CTR for Variant B
avg_ctr_variant_b = ctr_variant_b['CTR'].mean()
print("Average CTR for Variant B:", avg_ctr_variant_b)

Average CTR for Variant B: 0.1089330399532712


In [27]:
#Time Spent AVG & CTR for Varinat B
ctr_variant_b.mean()

/var/folders/fm/9pz_54hn7ynbrz17999wv77m0000gn/T/ipykernel_24386/1182103300.py:2: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  ctr_variant_b.mean()


Age           41.466667
Time Spent    23.343783
CTR            0.108933
dtype: float64

As we can see,Varinat A‘s average CTR is 0.120269,and average time spent is 24.805547.Also,Varinat B‘s average CTR is 0.108933,and average time spent is 23.343783. Thus,the Variant A of average CTR is higher than Variant B, so Variant A results in CTR or Time Spent lift.

Question2：Conduct statistical testing to determine if there is a statistically significant difference between the features and the control group.

Let's do an t-test next for each variant compared to the control, and then the variants compared to each other.

Null hypothesis: There is no significant difference between the CTR of the control group and Variant A or B
Alternative hypothesis: There is a significant difference between the CTR of the control group and Variant A or B
Indicates whether the null hypothesis should be rejected or rejected based on the significance level (alpha = 0.05).

In [62]:
# checking CTR
var_a = df[df['Variant']=='Variant A']['CTR']
var_b = df[df['Variant']=='Variant B']['CTR']
control = df[df['Variant']=='Control']['CTR']

t_stat,p_val = stats.ttest_ind(var_a, control)
t_stat_2,p_val_2 = stats.ttest_ind(var_b, control)
t_stat_3,p_val_3 = stats.ttest_ind(var_a, var_b)

print(f"Ttest_indResult: t-statistic = {t_stat:.4f}, p-value = {p_val:.4f}")
print(f"Ttest_indResult: t-statistic = {t_stat_2:.4f}, p-value = {p_val_2:.4f}")
print(f"Ttest_indResult: t-statistic = {t_stat_3:.4f}, p-value = {p_val_3:.4f}")

Ttest_indResult: t-statistic = 13.8294, p-value = 0.0000
Ttest_indResult: t-statistic = 6.4718, p-value = 0.0000
Ttest_indResult: t-statistic = 7.0850, p-value = 0.0000


First, Variant A & Control:
Reject the null hypothesis. There is a significant difference between the CTR of the control group and Variant A. The CTR of Variant A is significantly different from the control group.
Also,Variant B & Control:
We reject the null hypothesis. There is a significant difference between the CTR of the control group and Variant B. The CTR of Variant B is significantly different from the control group.
Then,Variant A & Variant B:
We reject the null hypothesis. There is a significant difference between the CTR of Variant A and Variant B. The CTR of Variant A is significantly different from the CTR of Variant B.

Null hypothesis: There is no significant difference between the Time Spent of the control group and Variant A or B.
Alternative hypothesis: There is a significant difference between the Time Spent of the control group and Variant A or B.

In [61]:
# checking Time Spent
var_a = df[df['Variant']=='Variant A']['Time Spent']
var_b = df[df['Variant']=='Variant B']['Time Spent']
control = df[df['Variant']=='Control']['Time Spent']

t_stat,p_val = stats.ttest_ind(var_a, control)
t_stat_2,p_val_2 = stats.ttest_ind(var_b, control)
t_stat_3,p_val_3 = stats.ttest_ind(var_a, var_b)

print(f"Ttest_indResult: t-statistic = {t_stat:.4f}, p-value = {p_val:.4f}")
print(f"Ttest_indResult: t-statistic = {t_stat_2:.4f}, p-value = {p_val_2:.4f}")
print(f"Ttest_indResult: t-statistic = {t_stat_3:.4f}, p-value = {p_val_3:.4f}")

Ttest_indResult: t-statistic = 12.1424, p-value = 0.0000
Ttest_indResult: t-statistic = 8.1742, p-value = 0.0000
Ttest_indResult: t-statistic = 3.6788, p-value = 0.0003


Both Variant A and Variant B show a statistically significant difference in average time spent compared to the control group. Additionally, there is a significant difference in average time spent between Variant A and Variant B.

In [32]:
# Three variant categories 
control_group = df[df['Variant'] == 'Control']
a_variant = df[df['Variant'] == 'Variant A']
b_variant = df[df['Variant'] == 'Variant B']

In [33]:
# Perform t-test for avg CTR
ctr_control_group = control_group['CTR']
ctr_a_variant = a_variant['CTR']
ctr_b_variant = b_variant['CTR']

In [34]:
# t-test bewteen control and variant A
t_stat_a, p_value_a = ttest_ind(ctr_control_group, ctr_a_variant)
print(f"T-statistic: {t_stat_a:.5f}")
print(f"P-value: {p_value_a:.5f}")


T-statistic: -13.82942
P-value: 0.00000


In [35]:
# t-test bewteen control and variant B
t_stat_b, p_value_b = ttest_ind(ctr_control_group, ctr_b_variant)
print(f"T-statistic: {t_stat_b:.5f}")
print(f"P-value: {p_value_b:.5f}")

T-statistic: -6.47181
P-value: 0.00000


Let's use an ANOVA test to compare all 3 groups and also t-tests to compare each variants to the Control, and to each other.

In [74]:
# checking CTR
var_A = df[df['Variant']=='Variant A']['CTR']
var_B = df[df['Variant']=='Variant B']['CTR']
control = df[df['Variant']=='Control']['CTR']

f_stat, p_value = stats.f_oneway(var_A, var_B, control)
print(f"F-statistic: {f_stat:.5f}")
print(f"P-value: {p_value:.5f}")

F-statistic: 93.58892
P-value: 0.00000


In [75]:
# checking Time Spent
var_A = df[df['Variant']=='Variant A']['Time Spent']
var_B = df[df['Variant']=='Variant B']['Time Spent']
control = df[df['Variant']=='Control']['Time Spent']

f_stat, p_value = stats.f_oneway(var_A, var_B, control)
print(f"F-statistic: {f_stat:.5f}")
print(f"P-value: {p_value:.5f}")


F-statistic: 75.60841
P-value: 0.00000


Based on these results, we can conclude that there are significant differences in CTR and average time spent between the control group, Variation A, and Variation B. This shows that the changes made in both variants had an impact on user engagement compared to the control group. Further analysis and exploration can be performed to determine the specific nature of these differences and determine which variant performs better in terms of user engagement.

Question3:Summarize your results. Make a recommendation to the engineering team about which feature to deploy. 

Conclusion: p-value is equal to 0 in these results compared between Variant A, B, and control group. There is a statistically significant difference in CTR between the control group and Variation A and Variation B. Also, there is a significant difference in CTR between Variation A and Variation B.

Recommendations:
1.Given the higher average CTR of Variant A compared to Variant B and the significant difference observed in the T-test, I would recommended to deploy Variant A as it appears to lead to an increase in CTR.
2.Further analysis and user feedback should be considered to understand the potential impact of variant A on other metrics and overall user experience before making a final decision.

Question4:Create a roll-out plan. How quickly will you introduce the feature to your audience?

When developing a rollout plan to introduce the feature to your audience, it is important to consider audience size, potential impact on user experience, and satisfaction and further improvement.
1. Soft launch: promote variant A to a small audience, such as 5-10%, for short-term testing and statistics within 1 week. Potential issues or bugs can be identified and fixed during the early development phase within a week. Receive customer satisfaction feedback and user experience, and improve in time.Monitor the value of CTR and keep it growing. This helps minimize negative impact on user experience and reduces the risk of widespread issues affecting a large user base.
2. Evaluate and test in the third week, analyze the characteristics of Variant A's audience for CTR(10%-35%), each stage allows the evaluation and analysis of key indicators, such as CTR, time spent and revenue. Identify trends in customer preferences, understand the The impact of functions on both sides of the user, and iterative improvement of the product as needed. This iterative process helps improve small issues of detail and usability.
3. Approximate customer target groups(35%-65%) and CTR changes for Variant A after one month. In addition to quantitative metrics, gathering qualitative user feedback is key to future impact. Gain insight into user satisfaction, perception, and any discomfort they may experience. This process can aid in decision-making and refinement scenarios at each later stage of rollout.
4. In the scalability and performance evaluation stage, after 3 months, gradually increase the exposure of this function to a larger proportion of the audience (65%-80%), allowing for product scalability and higher performance evaluation. This approach helps measure the impact and issues of the feature as the user base grows, ensures that the approximate functionality meets the needs of the user base, and ensures that the feature remains applicable across different usage scenarios.
5. When the function develops to a plateau stage (6 months, 90%), conduct profit evaluation, in addition to CTR and Time Spent, and track the impact of revenue goals and financial impact on function development. This includes average transaction value or any other revenue-related metric. Following the income indicator helps determine if the feature is positively impacting the economy and planning for the future